packages/strapi-admin/services/permission/permissions-manager/query-builers.js

```diff
@@ -98,7 +46,7 @@ const flattenDeep = condition => {
         set(...getTransformParams(key, v, k));
       });
     } else {
-      set(key, value);
+      set(key, flattenDeep(value));
```

Comment:
This recursion is here because **we need to handle cases where value is an array of non-primitives values** which need to be transformed as well.

This scenario is possible due to the previous conditions of the if/else if statement (we explicitly say that we must ignore arrays in the `else if` (which would cause issues on the flattening process such as `$or: [2]` => `_or.0: 2`))

By doing a recursion on the value here, we handle both the case where we have an array of primitives values, a mix of primitive and nonprimitive values (which.. well shouldn't happen), and only nonprimitives values.

```
if input is array => recurseMap(input)
if input is not object => input
else => do the flattening
```

Example of a scenario where the bugs can happen:

``` javascript
{
  $or: [
    { a: 2 },
    [
      { a: 3 },
      {
        $or: [
          { b: 1 }, 
          { b: 4 }
        ]
      }
    ]
  ]
}
```

would become

``` javascript
{
  _or: [
    { a: 2 },
    [
      { a: 3 },
      {
        $or: [
          { b: 1 },
          { b: 4 }
        ]
      }
    ]
  ]
}
```

if we don't apply the recursion in the else statement. The reason is that the second element of the value of the first `$or` would be simply put in the result as `key: value` without transformation.

By applying the recursion in the else, we end up with:

``` javascript
{
  _or: [
    { a: 2 },
    [
      { a: 3 },
      {
        _or: [
          { b: 1 },
          { b: 4 }
        ]
      }
    ]
  ]
}
```

which is the expected result.

Let me know if something wasn't clear enough.